optimizely/notification_center.py

```diff
@@ -14,43 +14,52 @@
 from functools import reduce
 
 from .helpers import enums
+from . import logger as optimizely_logger
+
+
+NOTIFICATION_TYPES = tuple(getattr(enums.NotificationTypes, attr)
+                           for attr in dir(enums.NotificationTypes)
+                           if not attr.startswith('__'))
 
 
 class NotificationCenter(object):
-  """ Class encapsulating Broadcast Notifications. The enums.NotifcationTypes includes predefined notifications."""
+  """ Class encapsulating methods to manage notifications and their listeners.
+  The enums.NotificationTypes includes predefined notifications."""
 
-  def __init__(self, logger):
-    self.notification_id = 1
-    self.notifications = {}
-    for (attr, value) in enums.NotificationTypes.__dict__.items():
-      self.notifications[value] = []
-    self.logger = logger
+  def __init__(self, logger=None):
+    self.listener_id = 1
+    self.notification_listeners = {}
+    for notification_type in NOTIFICATION_TYPES:
+      self.notification_listeners[notification_type] = []
+    self.logger = optimizely_logger.adapt_logger(logger or optimizely_logger.NoOpLogger())
 
   def add_notification_listener(self, notification_type, notification_callback):
-    """ Add a notification callback to the notification center.
+    """ Add a notification callback to the notification center for a given notification type.
 
     Args:
-      notification_type: A string representing the notification type from .helpers.enums.NotificationTypes
-      notification_callback: closure of function to call when event is triggered.
+      notification_type: A string representing the notification type from helpers.enums.NotificationTypes
+      notification_callback: Closure of function to call when event is triggered.
 
     Returns:
-      Integer notification id used to remove the notification or -1 if the notification has already been added.
+      Integer notification ID used to remove the notification or
+      -1 if the notification listener has already been added or
+      if the notification type is invalid.
     """
 
-    if notification_type not in self.notifications:
-      self.notifications[notification_type] = [(self.notification_id, notification_callback)]
-    else:
-      if reduce(lambda a, b: a + 1,
-                filter(lambda tup: tup[1] == notification_callback, self.notifications[notification_type]),
-                0) > 0:
-          return -1
-      self.notifications[notification_type].append((self.notification_id, notification_callback))
+    if notification_type not in NOTIFICATION_TYPES:
+      self.logger.error('Invalid notification_type: {} provided. Not adding listener.'.format(notification_type))
+      return -1
 
-    ret_val = self.notification_id
+    if reduce(lambda a, b: a + 1, filter(lambda tup: tup[1] == notification_callback,
```

Comment:
I don't find this to be very readable, but then again, I am not fluent in Python :)

Reply:
Let me try and update this as well. This entire thing was not very readable.